### PR TITLE
Filesystem fixes

### PR DESCRIPTION
**Pull request by NicknineTheEagle**
* Fix floppy mount label not updating on rescan. Was broken by an optimization.
* Return error in local FS on file write failure. Local FS currently returns success even on complete write failure.
* Added a setting to config that allows user to allow/disallow r/w opens of write-protected files in local FS. This makes behavior added in https://github.com/dosbox-staging/dosbox-staging/pull/143 optional - while it's more user-friendly it's also inaccurate to real DOS so it should be possible to turn it off.

### CODE REVIEW

File: src/dos/drive_cache.cpp [hunk 1]
```diff
@@ -142,11 +142,10 @@ void DOS_Drive_Cache::SetBaseDir(const char *baseDir)
 		return;
 
 	// Guard if source and destination are the same
-	if (basePath == baseDir) {
-		return;
+	if (basePath != baseDir) {
+		safe_strcpy(basePath, baseDir);
 	}
```
Comment (kcgen) [inline src/dos/drive_cache.cpp:148]: Perfect fix :+1:

File: src/dos/drive_local.cpp [hunk 2]
```diff
@@ -713,6 +713,12 @@ bool localFile::Write(uint8_t *data, uint16_t *size)
 	if (actual != requested) {
 		DEBUG_LOG_MSG("FS: Only wrote %u of %u requested bytes to file %s",
 		              actual, requested, name.c_str());
+
+		// Check for host write error
+		if (ferror(fhandle)) {
+			DOS_SetError(DOSERR_ACCESS_DENIED);
+			return false;
+		}
```
Comment (kcgen) [inline src/dos/drive_local.cpp:731]: Nice addition!

File: src/dos/drive_local.cpp [hunk 3]
```diff
@@ -154,9 +154,15 @@ bool localDrive::FileOpen(DOS_File **file, char *name, uint32_t flags)
 		// If yes, check if the file can be opened with Read-only access:
 		fhandle = fopen_wrap(newname, "rb");
 		if (fhandle) {
+			if (!always_allow_rw_open)
+				fhandle = NULL;
 
 #ifdef DEBUG
-			open_msg = "wanted writes but opened read-only";
+			if (can_rw_open_protected_files) {
```
Comment (kcgen) [inline src/dos/drive_local.cpp:161]: Maybe `can_rw_open_protected_files` was a previous name for `always_allow_rw_open`?

File: src/dos/drive_local.cpp [hunk 4]
```diff
@@ -154,9 +154,15 @@ bool localDrive::FileOpen(DOS_File **file, char *name, uint32_t flags)
 		// If yes, check if the file can be opened with Read-only access:
 		fhandle = fopen_wrap(newname, "rb");
 		if (fhandle) {
+			if (!always_allow_rw_open)
+				fhandle = NULL;
```
Comment (kcgen) [inline src/dos/drive_local.cpp:158]: In >= C++11 `nullptr` instead of `NULL` (because `nullptr` is a strict type for a pointer to memory, where as `NULL` is a basic integer) https://stackoverflow.com/a/14112019

Comment (kcgen) [reply src/dos/drive_local.cpp:158]: Also - I like putting the check at this level because it still lets the debug-mode messages trigger (instead of adding the check earlier).

File: src/dos/drive_local.cpp (hunk 3 above)
Comment (NicknineTheEagle) [reply src/dos/drive_local.cpp:161]: Wanted to make internal var name more verbose while keeping config var name concise. Still not sure about these, any alternative name suggestions are welcome.

File: src/dos/drive_local.cpp (hunk 4 above)
Comment (NicknineTheEagle) [reply src/dos/drive_local.cpp:158]: Oh, right, I normally use nullptr in Staging, just forgot this time.

File: src/dosbox.cpp [hunk 5]
```diff
@@ -581,6 +581,13 @@ void DOSBOX_Init()
 	        "  quiet       |   no    |    no\n"
 	        "  auto        | 'low' if exec or dir is passed, otherwise 'high'");
 
+	Pbool = secprop->Add_bool("always_allow_rw_open", only_at_start, true);
+	Pbool->Set_help(
+	        "Allow host write-protected data files to be read when opened for read-write by\n"
+	        "the DOS program. This avoids generating a false write operation OverlayFS,\n"
+	        "copy-on-write, and network filesystems when the data file(-s) are known to be\n"
+	        "read-only.");
```
Comment (kcgen) [inline src/dosbox.cpp:589]: This is a great and concise description of how it works, @NicknineTheEagle.  For technical users, they can read this and figure out the use-cases.

The benefits can be elusive though (until you actually work with it and see it in action) - I was thinking we should try to describe it from the user-case (or benefits) perspective. 

Some wording here:

*"Many games open all their files with writable permissions; even files that they never modify. This setting lets you write-protect those files while still allowing the game to read them. A second use-case: if you're using a copy-on-write or network-based filesystem, this setting  avoids triggering write-operations for these write-protected files."*

Unfortunately it's a bit wordly, and not sure how to tighten it up. Help wanted.

@NicknineTheEagle , @johnnovak , @FeralChild64 : what do you think?

File: src/dos/drive_local.cpp (hunk 3 above)
Comment (kcgen) [reply src/dos/drive_local.cpp:161]: Oh, what I meant to say is that `can_rw_open_protected_files` doesn't exist as a variable in the code - so this `DEBUG`-enabled will fail to compile.

It should be replaced with your current `always_allow_rw_open` bool, which I think is nicely self-documenting! :+1:

File: src/dos/drive_local.cpp (hunk 4 above)
Comment (kcgen) [reply src/dos/drive_local.cpp:158]: np!

File: src/dos/program_mount.cpp [hunk 6]
```diff
@@ -374,7 +378,14 @@ void MOUNT::Run(void) {
...
 			} else {
-				newdrive = new localDrive(temp_line.c_str(),sizes[0],int8_tize,sizes[2],sizes[3],mediaid);
+				newdrive = new localDrive(
+				        temp_line.c_str(),
+				        sizes[0],
+				        int8_tize,
+				        sizes[2],
+				        sizes[3],
+				        mediaid,
+				        section->Get_bool("always_allow_rw_open"));
```
Comment (kcgen) [inline src/dos/program_mount.cpp:388]: The feature is all about letting users write-protect data files (or setting read-only protection), so I think the conf setting should use that kind of wording; ie: "readonly", "write-protected", etc, instead of "rw".

Comment (NicknineTheEagle) [reply src/dos/program_mount.cpp:388]: You have to be mindful of var name lengths here because they affect column width in the config file.

Comment (kcgen) [reply src/dos/program_mount.cpp:388]: Yes, just thinking that we should avoid terms like "rw" and try to use "ro". 

Currently the users has to do some mental gymnastics to figure out the implication of "always_allow_rw_open" is thus to allow the use of write-protected or read-only files. 

(I unfortunately don't anything to offer as an alternative :sweat_smile: !)

Comment (johnnovak) [reply src/dos/program_mount.cpp:388]: We use longer config names than this, and I think having descriptive and discoverable names trumps all other concerns. E.g we have `viewport_resolution`, not `viewport_res` for example, plus check out the mouse related names. The column width is not a problem in the grand scheme of things.

I agree with @kcgen on this, it should have one of the terms he listed in the name. I'll think about it.

File: src/dosbox.cpp (hunk 5 above)
Comment (johnnovak) [reply src/dosbox.cpp:589]: I like @kcgen's version and I don't think it's too wordy. The original description describes the mechanics of the feature accurately and goes into technical details, but doesn't quite make it obvious why would I want to use this feature.

File: src/dos/program_mount.cpp (hunk 6 above)
Comment (johnnovak) [reply src/dos/program_mount.cpp:388]: I think I like `allow_write_protected_files`; this name and @kcgen's revised description should make things clear. Thought?

Comment (kcgen) [reply src/dos/program_mount.cpp:388]: For users interested in this feature, I think that really helps, @johnnovak. 

If a user sets `allow_write_protected_files = false` but /has/ write-protected their games' data files, then this setting will make sense when these problematic games are denied access to these files.

Comment (kcgen) [reply src/dos/program_mount.cpp:388]: >  I think having descriptive and discoverable names trumps all other concerns

yup - especially settings like this, which are typically going to be change while the user is editing their conf text files.

(as opposed to on-the-fly command-line changes, like switching sbtype from sb16 to sbpro2).

Comment (johnnovak) [reply src/dos/program_mount.cpp:388]: > > I think having descriptive and discoverable names trumps all other concerns
> 
> yup - especially settings like this. These are going to be "one-and-done", weighed against their personal workflow, while editing their conf text file.
> 
> As opposed to "tactical" changes on the command-line, like switching `sbtype` from `sb16` to `sbpro2` for some side-by-side comparisons.

Yeah I agree, and as one of my previous colleagues put it: as a developer, you're always optimising for _something_—whether consciously or out of sheer habit.

I'd very much prefer if we made these decisions consciously and based on things that really make a difference. For example, at the code level I'd say readability is of primary concern, and we should always optimise for that and easy maintainability. Due to other considerations (e.g. performance) we might relax that approach when necessary, but those special cases need to be well-justified and backed up with facts and numbers—otherwise we should default to well-structured code with descriptive variable and function names, etc.

Similarly for config parameters, I want all these parameter names to be as obvious as possible. `allow_write_protected_files` is 27 chars long, a bit longer than the other descriptive names we've added recently, but it describes the intent well. As a general rule, we could agree on keeping the names below 30-32 chars perhaps. 

For reference, these are some of the more recently added config params that all have good names and are not on th short side:

`mouse_middle_release`
`dos_mouse_immediate`
`ps2_mouse_model`
`sb_filter_always_on`
`ps1audio_dac_filter`
`viewport_resolution`
`pause_when_inactive`
`mcb_fault_strategy`

Comment (johnnovak) [reply src/dos/program_mount.cpp:388]: Also, I'd like to point out that the description of the body text needs to be wrapped to 79 chars, indeed, but the length of the config params doesn't matter when you use the `config -h <paramname>` command:

image

When you view the autogenrated config file in an editor, the line-lengths are all over the place already anyway, and this is "unfixable", so I wouldn't worry about it at all. Just use a big enough window as shown on the screenshot, and problem solved. In the grand scheme of things, this is a very minor concern. I do think trying to enforce 78-80 columns in the generated config is pointless—even terminal users can simply enlarge their terminal window these days 😄 

image

Comment (kcgen) [reply src/dos/program_mount.cpp:388]: All very well said @johnnovak !
This is great info that would fit well in the maintenance wiki page (or contributors docs!)

File: include/drives.h [hunk 7]
```diff
@@ -94,6 +97,7 @@ class localDrive : public DOS_Drive {
 
 private:
 	bool IsFirstEncounter(const std::string& filename);
+	bool always_allow_rw_open;
```
Comment (kcgen) [inline include/drives.h:100]: Suggest aligning this variable with the conf name, maybe `bool always_open_protected_files = `?

Comment (NicknineTheEagle) [reply include/drives.h:100]: Way too long + current name is more descriptive of the actual internal functionality.

Comment (kcgen) [reply include/drives.h:100]: `always_open_ro_files`

**Review by kcgen — APPROVED** [review]

**Comment by kcgen** [comment]
Thanks for yet another round of nice fixes, @NicknineTheEagle !
